src/util/mpsc/batch.rs

```diff
@@ -0,0 +1,307 @@
+    notifier_registered: AtomicBool,
+}
+
+impl State {
+    fn new(notify_size: usize) -> State {
+        State {
+            recv_task: AtomicOption::new(),
+            notify_size,
+            old_pending: AtomicUsize::new(0),
+            notifier_registered: AtomicBool::new(false),
+        }
+    }
+
+    #[inline]
+    fn inc_send_and_maybe_notify(&self) {
+        let old_pending = self.old_pending.fetch_add(1, Ordering::SeqCst);
+        if old_pending >= self.notify_size - 1 {
+            self.notify();
+        }
+    }
+
+    #[inline]
+    fn notify(&self) {
+        let task = self.recv_task.take(Ordering::AcqRel);
+        if let Some(t) = task {
+            self.old_pending.store(0, Ordering::SeqCst);
+            t.notify();
+        }
+    }
+}
+
+/// `Notifier` is used to notify receiver whenever you want.
+pub struct Notifier(Arc<State>);
+impl Notifier {
+    #[inline]
+    pub fn notify(self) {
+        drop(self);
+    }
+}
+
+impl Drop for Notifier {
+    #[inline]
+    fn drop(&mut self) {
+        let notifier_registered = &self.0.notifier_registered;
+        if !notifier_registered.compare_and_swap(true, false, Ordering::SeqCst) {
+            unreachable!("notifier_registered must be true");
+        }
+        self.0.notify();
+    }
+}
+
+pub struct Sender<T> {
+    sender: crossbeam_channel::Sender<T>,
+    state: Arc<State>,
+}
+
+impl<T> Clone for Sender<T> {
+    #[inline]
+    fn clone(&self) -> Sender<T> {
+        Sender {
+            sender: self.sender.clone(),
+            state: Arc::clone(&self.state),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    #[inline]
+    fn drop(&mut self) {
+        self.state.notify();
+    }
+}
+
+pub struct Receiver<T> {
+    receiver: crossbeam_channel::Receiver<T>,
+    state: Arc<State>,
+}
+
+impl<T> Sender<T> {
+    #[inline]
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        self.sender.send(t)?;
+        self.state.inc_send_and_maybe_notify();
+        Ok(())
+    }
+
+    #[inline]
+    pub fn try_send(&self, t: T) -> Result<(), TrySendError<T>> {
+        self.sender.try_send(t)?;
+        self.state.inc_send_and_maybe_notify();
+        Ok(())
+    }
+
+    #[inline]
+    pub fn get_notifier(&self) -> Option<Notifier> {
+        let notifier_registered = &self.state.notifier_registered;
+        if !notifier_registered.compare_and_swap(false, true, Ordering::SeqCst) {
+            return Some(Notifier(Arc::clone(&self.state)));
+        }
+        None
+    }
+}
+
+impl<T> Receiver<T> {
+    #[inline]
+    pub fn recv(&self) -> Result<T, RecvError> {
+        self.receiver.recv()
+    }
+
+    #[inline]
+    pub fn try_recv(&self) -> Result<T, TryRecvError> {
+        self.receiver.try_recv()
+    }
+
+    #[inline]
+    pub fn recv_timeout(&self, timeout: Duration) -> Result<T, RecvTimeoutError> {
+        self.receiver.recv_timeout(timeout)
+    }
+
+    #[inline]
+    fn yield_poll(&self) -> bool {
```

Comment:
what does yield mean?

Reply:
Just like yield instruction of CPU, or yield keywork in Python. Here it should means the `Receiver` will give up CPU and let Executors execute other futures.